Source/ASRunLoopQueue.mm

```diff
@@ -464,3 +484,240 @@ - (void)unlock
+#endif
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  CFRunLoopRemoveSource(_runLoop, _runLoopSource, kCFRunLoopCommonModes);
+  CFRelease(_runLoopSource);
+  _runLoopSource = nil;
+
+  if (CFRunLoopObserverIsValid(_preTransactionObserver)) {
+    CFRunLoopObserverInvalidate(_preTransactionObserver);
+  }
+  if (CFRunLoopObserverIsValid(_postTransactionObserver)) {
+    CFRunLoopObserverInvalidate(_postTransactionObserver);
+  }
+  CFRelease(_preTransactionObserver);
+  CFRelease(_postTransactionObserver);
+  _preTransactionObserver = nil;
+  _postTransactionObserver = nil;
+}
+
+#if ASRunLoopQueueLoggingEnabled
+- (void)checkRunLoop
+{
+  NSLog(@"<%@> - Jobs: %ld", self, _internalQueue.size());
+}
+#endif
+
+- (void)processQueue
+{
+  // If we have an execution block, this vector will be populated, otherwise remains empty.
+  // This is to avoid needlessly retaining/releasing the objects if we don't have a block.
+  std::vector<id> itemsToProcess;
+
+  {
+    ASDN::MutexLocker l(_internalQueueLock);
+
+    // Mark the queue will end coalescing shortly until after CATransactionCommit.
+    // This will give the queue a chance to apply any further interfaceState changes/enqueue
+    // immediately within current runloop instead of pushing the work to next runloop cycle.
+    _CATransactionCommitInProgress = YES;
+
+    NSInteger internalQueueCount = _internalQueue.count;
+    // Early-exit if the queue is empty.
+    if (internalQueueCount == 0) {
+      return;
+    }
+
+    ASSignpostStart(ASSignpostRunLoopQueueBatch);
+
+    /**
+     * For each item in the next batch, if it's non-nil then NULL it out
+     * and if we have an execution block then add it in.
+     * This could be written a bunch of different ways but
+     * this particular one nicely balances readability, safety, and efficiency.
+     */
+    NSInteger foundItemCount = 0;
+    for (NSInteger i = 0; i < internalQueueCount && foundItemCount < internalQueueCount; i++) {
+      /**
+       * It is safe to use unsafe_unretained here. If the queue is weak, the
+       * object will be added to the autorelease pool. If the queue is strong,
+       * it will retain the object until we transfer it (retain it) in itemsToProcess.
+       */
+      __unsafe_unretained id ptr = (__bridge id)[_internalQueue pointerAtIndex:i];
+      if (ptr != nil) {
+        foundItemCount++;
+        itemsToProcess.push_back(ptr);
+        [_internalQueue replacePointerAtIndex:i withPointer:NULL];
+      }
+    }
+
+    if (foundItemCount == 0) {
+      // If _internalQueue holds weak references, and all of them just become NULL, then the array
+      // is never marked as needsCompletion, and compact will return early, not removing the NULL's.
+      // Inserting a NULL here ensures the compaction will take place.
+      // See http://www.openradar.me/15396578 and https://stackoverflow.com/a/40274426/1136669
+      [_internalQueue addPointer:NULL];
+    }
```

Comment:
If the queue only holds strong references, this trick shouldn't be necessary.

Reply:
@cesteban awesome catch, thanks for reviewing! @wsdwsd0829 this would be a good change to make, since we expect this queue to always be strong (otherwise the node could dealloc before getting the invisible state) - so I think we can just delete these few lines.

Reply:
Thanks, fixed.